draft of repo and readme (#1)
ocaml/opam-repository-archive: README.md

# Opam-repository-archive

This repository contains archived OCaml packages, taken from the [main opam-repository](https://github.com/ocaml/opam-repository).

## Motivation

The motivation is that the [main opam-repository](https://github.com/ocaml/opam-repository) contains lots of packages, and lots of versions of packages. And a large part of older versions, patch releases, etc. - are no longer maintained or used. With the biannual releases of OCaml, there's also only a limited amount of OCaml versions supported by the main opam-repository (reducing CPU cycles).

> The most important reason for the existance of this repository is to remove a lot of burden (waiting for the computer) for OCaml developers that are not interested in 95% percent of the old opam package releases - and given the design of opam (text files, SMT solver), its performance is not linear with the amount of opam files in the repository.

Reviewer: ```suggestion
The most important reason for the existence of this repository is to reduce burden for OCaml developers that are not interested in many of the old opam package releases -- and given the design of opam (text files, SMT solver), its performance is not linear with the amount of opam files in the repository -- and to improve the experience for new users in search of the right libraries -- by clearing up the space from broken and unmaintained packages.
```

This is the home for packages that are not very likely to be of interest for the common OCaml developer. But when hunting a bug or performance regression or doing archeology, these packages may be very useful for you. Also, when doing changes to the OCaml compiler or runtime and wanting to do some quantative study.

If you're interested in more discussion on how this repository came to be there, read the [initial issue: How does opam-repository scale?](https://github.com/ocaml/opam-repository/issues/23789).

This is a medium-term solution for the question how to scale opam-repository, with the neat property that it works with deployed opam versions out of the box - no need to force everybody to have a new opam version which avoids the scaling issue by other means.

## Usage

You can add this repository to your active opam switch:

> `opam repo add archive git+https://github.com/ocaml/opam-repository-archive`

Reviewer: ```suggestion

    opam repo add archive git+https://github.com/ocaml/opam-repository-archive`

```

I would make the block separate for this

> If you want to add it to all opam switches, add `--all`, and `--set-default` to add it to all newly created switches as well.

Reviewer: ```suggestion
If you want to add it to all opam switches, add `--all` to the command above, and with `--set-default` you can add the repository to all newly created switches as well.
```

Afterwards you can install the available packages using `opam install mypackage` (or `mypackage.version`).

## Additional non-standard fields in the opam files

Each opam file in this repository contains some fields:

- `x-reason-for-archival`
  - Allowed values: A list of strings matching the primary repo criteria IDs (see below).
  - Meaning: Records the unmet primary repo criteria motivating the archiving.
  - Example: ["ocaml-version", "maintenance-intent"]
- `x-opam-repository-commit-hash-at-time-of-archival`
  - Allowed values: a string
  - Meaning: Records the commit hash of the primary repo at the time a package version is archived.
  - Example: "ca32ab3b976aa7abc00c7605548f78a30980d35b"

Criteria ID:
- `ocaml-version`: The package version satisfies the current compiler cutoff threshold.

> - `source-available`: The sources of the package version are available.

Reviewer: ```suggestion
- `source-unavailable`: The sources of the package version are unavailable.
```
I thought we wanted to use this intent to say that the package was archived since the sources were not available, maybe it would be better to rename it in that way

Author: These `Criteria ID` come from the package archiving policy. I have not been involved in developing these IDs, am happy to change whatever is suitable -- but I'd keep these in this README and those in the policy in sync.

So, it would be great if you @mseri @raphael-proust @shonfeder could decide on revising these IDs (or we can do that tomorrow in the opam-repo-meeting) and then move forward.

Reviewer 2: Rather than reiterating these in the readme here, could we not just refer to the policy document? Having a single source of truth for this kind of thing really seems preferable IMO.

Reviewer: Let's be explicit and say that these are the criteria to be in the opam-repository, and their appearence here means that they were no longer satisfied

Author: I negated the description in [11b9e76](https://github.com/ocaml/opam-repository-archive/pull/1/commits/11b9e7624eb40f9e1d8d80e779dbcdbae806919d)

# Opam-repository-archive

This repository contains archived OCaml packages, taken from the [main opam-repository](https://github.com/ocaml/opam-repository).

## Motivation

> The motivation is that the [main opam-repository](https://github.com/ocaml/opam-repository) contains lots of packages, and lots of versions of packages. And a large part of older versions, patch releases, etc. - are no longer maintained or used. With the biannual releases of OCaml, there's also only a limited amount of OCaml versions supported by the main opam-repository (reducing CPU cycles).

Author: removed in d997a18

> The most important reason for the existence of this repository is to reduce burden for OCaml developers that are not interested in many of the old opam package releases -- and given the design of opam (text files, SMT solver), its performance is not linear with the amount of opam files in the repository -- and to improve the experience for new users in search of the right libraries -- by clearing up the space from broken and unmaintained packages.

Author: rephrased differently in d997a18

This is the home for packages that are not very likely to be of interest for the common OCaml developer. But when hunting a bug or performance regression or doing archeology, these packages may be very useful for you. Also, when doing changes to the OCaml compiler or runtime and wanting to do some quantative study.

If you're interested in more discussion on how this repository came to be there, read the [initial issue: How does opam-repository scale?](https://github.com/ocaml/opam-repository/issues/23789).

> This is a medium-term solution for the question how to scale opam-repository, with the neat property that it works with deployed opam versions out of the box - no need to force everybody to have a new opam version which avoids the scaling issue by other means.

Author: I find this sentence here makes it clear that the opam-repository-archive is not a long-term solution, and there may be other paths in the future.

The discussion about "longer term" surely do not belong here.

## Usage

You can add this repository to your active opam switch:

    opam repo add archive git+https://github.com/ocaml/opam-repository-archive`

If you want to add it to all opam switches, add `--all` to the command above, and with `--set-default` you can add the repository to all newly created switches as well.

Afterwards you can install the available packages using `opam install mypackage` (or `mypackage.version`).

> ## Additional non-standard fields in the opam files

Author: From my perspective, having this README mostly self-contained makes it easy to comprehend. For me, the policy itself contains a lot of details and processes that are mainly for package authors and opam repository maintainers -- but this README is targeted at people who want to use some old package.
